doc/designs/quic-design/quic-io-arch.md

```diff
@@ -0,0 +1,485 @@
+  of the size we wish to send is writeable, so a send(3) call could block
+  anyway.
+
+- This usage pattern precludes multithreaded use barring some locking scheme
+  due to the possibility of other threads racing between the call to select(3)
+  and the subsequent I/O call. This undermines our intentions to support
+  multi-threaded network I/O on the backend.
+
+Moreover, our QUIC implementation will not drive the Berkeley sockets API
+directly but uses the BIO abstraction to access the network, so these issues are
+then compounded by the limitations of our existing BIO interfaces. We do not
+have a BIO interface which provides for select(3)-like functionality or which
+can implement the required semantics above. Therefore, trying to implement QUIC
+on top of blocking I/O in this way would require violating the BIO abstraction
+layer, and would not work with custom BIOs.
+
+### Blocking sockets and threads
+
+Another conceptual possibility is that blocking calls could be kept ongoing in
+parallel threads. Under this model, there would be three threads:
+
+- a thread which exists solely to execute blocking calls to the `BIO_write` of
+  an underlying network BIO,
+- a thread which exists solely to execute blocking calls to the `BIO_read` of an
+  underlying network BIO,
+- a thread which exists solely to wait for and dispatch timeout events.
+
+This has a large number of disadvantages:
+
+- There is a hard requirement for threading functionality in order to be
+  able to support blocking semantics at the application level. Applications
+  which require blocking semantics would only be able to function in thread
+  assisted mode. In environments where threading support is not available or
+  desired, our APIs would only be usable in a non-blocking fashion.
+
+- Several threads are spawned which the application is not in control of.
+  This undermines our general approach of providing the application with control
+  over OpenSSL's use of resources, such as allowing the application to do its
+  own polling or provide its own allocators.
+
+  At a minimum for a client, there must be two threads per connection. This
+  means if an application opens many outgoing connections, there will need
+  to be `2n` extra threads spawned.
+
+- By blocking in `BIO_write` calls, this precludes correct implementation of
+  QUIC. Unlike any analogue in TLS, QUIC packets are time sensitive and intended
+  to be transmitted as soon as they are generated. QUIC packets contain fields
+  such as the ACK Delay value, which is intended to describe the time between a
+  packet being received and a return packet being generated. Correct calculation
+  of this field is necessary to correct calculation of connection RTT. It is
+  therefore important to only generate packets when they are ready to be sent,
+  otherwise suboptimal performance will result. This is a usage model which
+  aligns optimally to non-blocking I/O and which cannot be accommodated
+  by blocking I/O.
```

Comment:
I'm not really convinced by this argument. If you consider the kernel's write buffers to be part of the "network" then this delay blocked on write is legitimately part of the RTT. I think that's a reasonable position to take - the "endpoint" is OpenSSL's QUIC implementation. Anything in between us and the other endpoint is "network".

Reply:
If the buffers are temporarily full this is an artifact of congestion, not the RTT... it is true this has historically been an issue where congestion generates *apparent* RTT increases, but this is not desirable and transport protocols are designed where possible to avoid it (see bufferbloat).